File: Ex2.cpp
#include <iostream>
#include <string>

using namespace std;

string decodeString(const string& s, size_t& i) {
    string result = "";

    while (i < s.length() && s[i] != ']') {
        if (!isdigit(s[i])) {
            result += s[i++];
        } 
        else {
            int count = 0;
            while (isdigit(s[i])) {
                count = count * 10 + (s[i++] - '0');
            }

            // Пропуск '['
            i++;
            string decodedString = decodeString(s, i);

            // Пропуск ']'
            i++;
            
            for (int j = 0; j < count; j++) {
                result += decodedString;
            }
        }
    }

    return result;
}

string decodeString(const string& s) {
    size_t i = 0;
    return decodeString(s, i);
}

int main() {
    string s = "3[a]2[bc]";
    cout << decodeString(s) << endl;

    s = "3[a2[c]]";
    cout << decodeString(s) << endl;

    s = "2[abc]3[cd]ef";
    cout << decodeString(s) << endl;

    return 0;
}

File: Ex2.py
def decode_string(s, i):
    result = []

    while i < len(s) and s[i] != ']':
        if not s[i].isdigit():
            result.append(s[i])
            i += 1
        else:
            count = 0
            while i < len(s) and s[i].isdigit():
                count = count * 10 + int(s[i])
                i += 1

            # Skip '['
            i += 1
            decoded_string, i = decode_string(s, i)

            # Skip ']'
            i += 1

            result.append(decoded_string * count)

    return ''.join(result), i

def decode_string_main(s):
    result, _ = decode_string(s, 0)
    return result

if __name__ == "__main__":
    s = "3[a]2[bc]"
    print(decode_string_main(s))

    s = "3[a2[c]]"
    print(decode_string_main(s))

    s = "2[abc]3[cd]ef"
    print(decode_string_main(s))
